kernel/inductive.ml

```diff
@@ -1430,16 +1482,18 @@ let inductive_of_mutfix env ((nvect,bodynum),(names,types,bodies as recdef)) =
             else anomaly ~label:"check_one_fix" (Pp.str "Bad occurrence of recursive call.")
         | _ -> raise_err env i NotEnoughAbstractionInFixBody
     in
-    let ((ind, _), _) as res = check_occur fixenv 1 def in
+    let ((ind, u), _) as res = check_occur fixenv 1 def in
     let _, mip = lookup_mind_specif env ind in
     (* recursive sprop means non record with projections -> squashed *)
-    if mip.mind_relevance == Sorts.Irrelevant &&
-       not (Environ.is_type_in_type env (GlobRef.IndRef ind))
-    then
-      begin
-        if not (names.(i).Context.binder_relevance == Sorts.Irrelevant)
-        then raise_err env i FixpointOnIrrelevantInductive
-      end;
+    let () =
+      if Environ.is_type_in_type env (GlobRef.IndRef ind) then ()
+      else match relevance_of_ind_body mip u with
+        | Sorts.Irrelevant | Sorts.RelevanceVar _ ->
+          (* XXX if RelevanceVar also allow binder_relevance = the same var? *)
+          if not (names.(i).Context.binder_relevance == Sorts.Irrelevant)
```

Comment:
This check is wrong. On non-ground instances the fix must have the same quality as the argument, because it's just a case in disguise.

Reply:
(For uniformity you should use `is_allowed_elimination`).

Reply:
I think we must allow elimination from `Sorts.RelevanceVar q` to `SProp` for both `Case` and `Fix` because otherwise it would not be possible to prove any properties on sort polymorphic lists. Note that this has two interesting consequences. The first one is that properties on sort polymorphic structures have to be stated in `SProp` and not in `Prop`. The second one is that only "positive" statement can be proven this way.  For instance, the fact that `nil` is not equal to `cons` in general cannot be proven because it requires large elimination, and thus elimination from `Sorts.RelevanceVar q` to `Type`. This is normal because no confusion is not valid for lists in `SProp`.

Reply:
changed

Reply:
actually not changed, I did regular eliminations but not this bit

Reply:
now really changed

Reply:
FTR we decided the other way around while talking offline with @tabareau for various reasons.